src/vsr/replica.zig

```diff
@@ -3691,11 +3573,215 @@ pub fn ReplicaType(
+                    self.commit_min,
+                });
+                return;
+            }
+
+            assert(self.commit_stage == .idle);
+            self.commit_dispatch_enter();
+        }
+
+        const CommitStage = union(enum) {
+            /// Not committing.
+            idle,
+            /// Get the next prepare to commit from the journal or pipeline,
+            start,
+            /// if there isn't one, break out of commit loop.
+            check_prepare,
+            /// Load required data from LSM tree on disk into memory.
+            prefetch,
+            /// Ensure that the ClientReplies has at least one Write available.
+            reply_ready,
+            /// Execute state machine logic.
+            execute,
+            /// Run one beat of LSM compaction
+            compact,
+            /// Every vsr_checkpoint_ops, persist the current state to disk,
+            checkpoint_data: struct {
+                done_set: std.enums.EnumSet(CheckpointData) = .{},
+
+                fn done(checkpoint_data: *@This(), stage: CheckpointData) bool {
+                    assert(!checkpoint_data.done_set.contains(stage));
+                    checkpoint_data.done_set.insert(stage);
+                    return checkpoint_data.done_set.count() ==
+                        std.enums.EnumSet(CheckpointData).len;
+                }
+            },
+            /// and update the superblock
+            checkpoint_superblock,
+
+            fn transition(stage_current: *CommitStage, stage_next: CommitStage) bool {
+                const valid = switch (stage_current.*) {
+                    .idle => stage_next == .start,
+                    .start => stage_next == .check_prepare,
+                    .check_prepare => stage_next == .prefetch,
+                    .prefetch => stage_next == .reply_ready,
+                    .reply_ready => stage_next == .execute,
+                    .execute => stage_next == .compact,
+                    .compact => stage_next == .checkpoint_data,
+                    .checkpoint_data => stage_next == .checkpoint_superblock,
+                    .checkpoint_superblock => stage_next == .idle,
+                };
+                if (!valid) return false;
+                stage_current.* = stage_next;
+                return true;
+            }
+        };
+
+        const CheckpointData = enum {
+            state_machine,
+            client_replies,
+            client_sessions,
+            grid,
+        };
+
+        /// The code bellow is a manual desugaring of loop with the following structure:
+        /// loop {
+        ///     prefetch().await;
+        ///     execute().await;
+        ///     compact().await;
+        /// }
+        fn commit_dispatch(self: *Self) void {
+            assert(!self.commit_dispatch_entered);
+            self.commit_dispatch_entered = true;
+            defer self.commit_dispatch_entered = false;
```

Comment:
Yeah, `commit_stage != .idle` is equivalent to your suggested `commit_dispatch_entered`. 

Though, I did follow your suggestion, because _not_ doing that actually results in failing asserts, because we do have re-entrancy here. When we fetch prepare from the journal, the callback can be called synchroneously if preapare doesn't have a body, or if it is missing. 

So both in the new code and in the old code, we _could_ enter `commit_dispatch` while already having a a running `commit_dispatch` on the call stack. I also want to fix that, but I am not sure I want to fix that in this PR